internal/id/id.go

```diff
@@ -1,38 +1,59 @@
 package id
 
 import (
-	"sync"
+	"sync/atomic"
 )
 
-var mu sync.Mutex
-var turn, cycle, counter uint32 = 0, 0, 0
-
-var totalBits uint32 = 32
-var turnBits uint32 = 4
-var counterBits = totalBits - turnBits
+const (
+	totalBits   uint32 = 32                                // Total bits for the ID
+	turnBits    uint32 = 4                                 // Bits for turn (allows up to 16 turns before wrap-around
+	workerBits  uint32 = 4                                 // Bits allocated for worker ID (supports up to 16 workers, 2^4)
+	counterBits uint32 = totalBits - workerBits - turnBits // Remaining bits for counter (24 bits)
+)
 
-var cycleMap []uint32
+var (
+	counter  []uint32        // Per-worker counter
+	turn     []uint32        // Per-worker turn counter
+	cycleMap [][]uint32      // Per-worker cycle map
+	maxTurns = 1 << turnBits // Max turns before wrap-around
+)
 
 func init() {
-	cycleMap = make([]uint32, 1<<turnBits)
-}
+	numWorkers := 1 << workerBits
+	counter = make([]uint32, numWorkers)
+	turn = make([]uint32, numWorkers)
+	cycleMap = make([][]uint32, numWorkers)
 
-func ExpandID(id uint32) uint64 {
-	_id := uint64(id)
-	_id |= uint64(cycleMap[id>>counterBits]) << counterBits
-	return _id
+	for i := range cycleMap {
+		cycleMap[i] = make([]uint32, maxTurns)
+	}
 }
 
-// NextID returns a new unique ID
-// TODO: Persisting the cycle on disk and reloading it when we start the server
-func NextID() uint32 {
-	mu.Lock()
-	defer mu.Unlock()
-	counter = (counter + 1) & ((1 << counterBits) - 1)
-	if counter == 0 {
-		cycle++
-		turn = (turn + 1) & ((1 << turnBits) - 1)
-		cycleMap[turn] = cycle
+// NextID generates a lock-free, unique ID per worker.
+// TODO: Persist the cycleMap to disk for server restart.
+func NextID(workerID uint32) uint32 {
+	if workerID >= (1 << workerBits) {
+		panic("workerID exceeds limit for allocated bits")
 	}
-	return (turn << counterBits) | counter
+
+	// Increment the counter for this worker and mask it to stay within limit
+	counterVal := atomic.AddUint32(&counter[workerID], 1) & ((1 << counterBits) - 1)
```

Comment:
What happens in case DiceDB restarts for some reason? cycleMap is stored in memory. Could this potentially leading to ID collisions if we are persisting the commands using WAL?